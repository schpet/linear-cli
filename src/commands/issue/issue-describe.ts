import { Command } from "@cliffy/command"
import { fetchIssueDetails, getIssueIdentifier } from "../../utils/linear.ts"
import { formatIssueDescription } from "../../utils/jj.ts"
import { getNoIssueFoundMessage } from "../../utils/vcs.ts"

export const describeCommand = new Command()
  .name("describe")
  .description("Print the issue title and Linear-issue trailer")
  .arguments("[issueId:string]")
  .action(async (_, issueId) => {
    const resolvedId = await getIssueIdentifier(issueId)
    if (!resolvedId) {
      console.error(getNoIssueFoundMessage())
      Deno.exit(1)
    }
    const { title, url } = await fetchIssueDetails(
      resolvedId,
      Deno.stdout.isTerminal(),
    )

    console.log(formatIssueDescription(resolvedId, title, url))
  })
