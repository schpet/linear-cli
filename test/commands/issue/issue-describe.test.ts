import { snapshotTest } from "@cliffy/testing"
import { describeCommand } from "../../../src/commands/issue/issue-describe.ts"
import { MockLinearServer } from "../../utils/mock_linear_server.ts"

// Common Deno args for permissions
const denoArgs = [
  "--allow-env=GITHUB_*,GH_*,LINEAR_*,NODE_ENV,EDITOR,SNAPSHOT_TEST_NAME",
  "--allow-read",
  "--allow-write",
  "--allow-run",
  "--allow-net",
  "--quiet",
]

// Test help output
await snapshotTest({
  name: "Issue Describe Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs,
  async fn() {
    await describeCommand.parse()
  },
})

// Test with working mock server
await snapshotTest({
  name: "Issue Describe Command - With Mock Server",
  meta: import.meta,
  colors: false,
  args: ["TEST-123"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetails",
        variables: { id: "TEST-123" },
        response: {
          data: {
            issue: {
              title: "Fix authentication bug in login flow",
              description:
                "Users are experiencing issues logging in when their session expires.",
              url:
                "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
              branchName: "fix/test-123-auth-bug",
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await describeCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test with issue not found
await snapshotTest({
  name: "Issue Describe Command - Issue Not Found",
  meta: import.meta,
  colors: false,
  args: ["TEST-999"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetails",
        variables: { id: "TEST-999" },
        response: {
          errors: [{
            message: "Issue not found: TEST-999",
            extensions: { code: "NOT_FOUND" },
          }],
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      try {
        await describeCommand.parse()
      } catch (error) {
        console.log(`Error: ${(error as Error).message}`)
      }
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})
