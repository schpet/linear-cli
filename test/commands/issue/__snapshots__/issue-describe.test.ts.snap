export const snapshot = {};

snapshot[`Issue Describe Command - Help Text 1`] = `
stdout:
"
Usage: describe [issueId]

Description:

  Print the issue title and Linear-issue trailer

Options:

  -h, --help  - Show this help.  

"
stderr:
""
`;

snapshot[`Issue Describe Command - With Mock Server 1`] = `
stdout:
"TEST-123 Fix authentication bug in login flow

Linear-issue: [TEST-123](https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow)
"
stderr:
""
`;

snapshot[`Issue Describe Command - Issue Not Found 1`] = `
stdout:
'Error: Issue not found: TEST-999: {"response":{"errors":[{"message":"Issue not found: TEST-999","extensions":{"code":"NOT_FOUND"}}],"status":200,"headers":{}},"request":{"query":"query GetIssueDetails(\$id: String!) {\\\\n  issue(id: \$id) {\\\\n    title\\\\n    description\\\\n    url\\\\n    branchName\\\\n  }\\\\n}","variables":{"id":"TEST-999"}}}
'
stderr:
"âœ— Failed to fetch issue details
"
`;
